src/v/archival/archiver_operations_impl.cc

```diff
@@ -61,12 +88,108 @@ class archiver_operations_impl : public archiver_operations_api {
+              partition->get_uploaded_offset());
+
+            // If not limit is specified (quota is set to nullopt) we don't want
+            // to upload unlimited amount of data in one iteration. The defaults
+            // in this case are configured to allow only three segments to be
+            // uploaded.
+            auto size_quota = arg.upload_size_quota.value_or(
+              arg.target_size * 3);
+            // up to 3 requests to upload one segment
+            constexpr size_t default_req_quota = 12;
+
+            auto req_quota = arg.upload_requests_quota.value_or(
+              default_req_quota);
+
+            size_limited_offset_range range(
+              base_offset, arg.target_size, arg.min_size);
+
+            retry_chain_node op_rtc(&workflow_rtc);
+
+            vlog(
+              _rtclog.debug,
+              "start collecting segments, base_offset {}, size quota {}, "
+              "requests "
+              "quota {}",
+              base_offset,
+              size_quota,
+              req_quota);
+
+            while (size_quota > 0 && req_quota > 0) {
+                auto upload = co_await make_non_compacted_upload(
+                  base_offset, partition, arg, _sg, op_rtc.get_deadline());
+
+                if (
+                  upload.has_error()
+                  && upload.error() == error_outcome::not_enough_data) {
+                    vlog(
+                      _rtclog.debug,
+                      "make_non_compacted_upload failed {}",
+                      upload.error());
+                    break;
+                }
+                if (upload.has_error()) {
+                    vlog(
+                      _rtclog.error,
+                      "make_non_compacted_upload failed {}",
+                      upload.error());
+                    co_return upload.error();
+                }
+                vlog(
+                  _rtclog.debug,
+                  "make_non_compacted_upload success {}",
+                  upload.value());
+                base_offset = model::next_offset(
+                  upload.value()->metadata.committed_offset);
+                size_quota -= upload.value()->size_bytes;
+                // 2 PUT requests for segment upload + index upload
+                req_quota -= 2;
+                if (upload.value()->metadata.metadata_size_hint > 0) {
+                    // another PUT request if tx-manifest is not empty
+                    req_quota -= 1;
+                }
+                result.results.push_back(std::move(upload.value()));
+            }
+
+            vlog(
+              _rtclog.debug,
+              "find_upload_candidates completed with {} results, read-write "
+              "fence "
+              "{}",
+              result.results.size(),
+              result.read_write_fence);
+            // TODO: compacted upload (if possible)
+            // TODO: merge adjacent segments
```

Comment:
Just noting that this seems like a great opportunity to decouple the non-compacted upload path from the compacted upload path. Are you sure we'd want to add these here? It will be much easier to reuse primitives that do fewer things, just thinking about when we get to different cloud storage implementations

Reply:
it's not implemented yet